python/ray/serve/pipeline/generate.py

```diff
@@ -0,0 +1,170 @@
+from typing import Any, Dict, List, OrderedDict
+import threading
+
+from ray.experimental.dag import (
+    DAGNode,
+    ClassNode,
+    ClassMethodNode,
+    FunctionNode,
+)
+from ray.serve.api import Deployment, DeploymentConfig
+from ray import serve
+from ray.serve.pipeline.deployment_method_node import DeploymentMethodNode
+from ray.serve.pipeline.deployment_node import DeploymentNode
+
+
+class DeploymentIDGenerator(object):
+    """
+    Generate unique suffix for each given deployment_name requested for id.
+    By default uses deployment_name for the very first time, then append
+    monotonic increasing id to it.
+    """
+
+    __lock = threading.Lock()
+    __shared_state = dict()
+
+    @classmethod
+    def get_deployment_id(cls, deployment_name: str):
+        with cls.__lock:
+            if deployment_name not in cls.__shared_state:
+                cls.__shared_state[deployment_name] = 0
+                return deployment_name
+            else:
+                suffix_num = cls.__shared_state[deployment_name] + 1
+                cls.__shared_state[deployment_name] = suffix_num
+
+                return f"{deployment_name}_{suffix_num}"
+
+def _get_unique_deployment_name(dag_node):
+    """
+    Get unique deployment name for given ClassNode. For the first time we use
+    class name, and subsequent instantition of the same class used in Ray DAG
+    get monotonic increasing id as suffix.
+    """
+    deployment_name = (
+        dag_node.get_options().get("name", None) or dag_node._body.__name__
+    )
+    deployment_name = DeploymentIDGenerator.get_deployment_id(
+        deployment_name
+    )
+    return deployment_name
+
+def _remove_non_default_ray_actor_options(ray_actor_options: Dict[str, Any]):
+    """
+    In Ray DAG building we pass full ray_actor_options regardless if a field
+    was explicitly set. Since some values are invalid, we need to remove them
+    from ray_actor_options.
+    """
```

Comment:
interestingly, there's a small edge case we likely won't handle: if the user passed in one of these default explicitly, we won't error but will instead just purge it. not important IMO but if there's a simple way to handle it, might be worthwhile

Reply:
i see, let me leave a TODO about this, we will revisit this part soon when we get to the `build()` implementations.